src/Language/Futhark/Interpreter/AD.hs

```diff
@@ -0,0 +1,289 @@
+import Data.Maybe (fromJust)
+import Data.Foldable (foldlM)
+
+data ADSeed
+  = VjpSeed Int Tape
+  | JvpSeed Int ADValue ADValue
+
+instance Show ADSeed where
+  show :: ADSeed -> String
+  show (VjpSeed d v) = "VjpSeed d" <> show d <> " " <> show v
+  show (JvpSeed d v dv) = "JvpSeed d" <> show d <> " " <> show v <> " " <> show dv
+
+data ADValue
+  = Primal PrimValue
+  | Seed ADSeed
+
+instance Show ADValue where
+  show (Primal v) = "Primal " <> show v
+  show (Seed s) = "Seed " <> show s
+
+data Op
+  = OpBin  BinOp
+  | OpCmp  CmpOp
+  | OpUn   UnOp
+  | OpFn   String
+  | OpConv ConvOp
+
+instance Show Op where
+  show (OpBin  op) = show op
+  show (OpCmp  op) = show op
+  show (OpUn   op) = show op
+  show (OpFn   fn) = show fn
+  show (OpConv op) = show op
+
+value :: ADValue -> PrimValue
+value (Primal v) = v
+value (Seed s) = value $ primal s
+
+depth :: ADSeed -> Int
+depth (VjpSeed d _) = d
+depth (JvpSeed d _ _) = d
+
+primal :: ADSeed -> ADValue
+primal (VjpSeed _ v) = tapeValue v
+primal (JvpSeed _ v _) = v
+
+valueAsType :: PrimValue -> Rational -> PrimValue
+valueAsType v c = case v of
+  IntValue (Int8Value _)      -> IntValue $ Int8Value  $ round c
+  IntValue (Int16Value _)     -> IntValue $ Int16Value $ round c
+  IntValue (Int32Value _)     -> IntValue $ Int32Value $ round c
+  IntValue (Int64Value _)     -> IntValue $ Int64Value $ round c
+  FloatValue (Float16Value _) -> FloatValue $ Float16Value $ fromRational c
+  FloatValue (Float32Value _) -> FloatValue $ Float32Value $ fromRational c
+  FloatValue (Float64Value _) -> FloatValue $ Float64Value $ fromRational c
+  _ -> error $ "No valid prim value for " ++ show v
+
+opTypeMatch :: Op -> [PrimValue] -> Bool
+opTypeMatch (OpBin  op) p = all (\x ->      binOpType  op  == primValueType x) p
+opTypeMatch (OpCmp  op) p = all (\x ->      cmpOpType  op  == primValueType x) p
+opTypeMatch (OpUn   op) p = all (\x ->      unOpType   op  == primValueType x) p
+opTypeMatch (OpConv op) p = all (\x -> fst (convOpType op) == primValueType x) p
+opTypeMatch (OpFn   fn) p = case M.lookup fn primFuns of
+                          Just (t, _, _) -> and $ zipWith (\x y -> x == primValueType y) t p
+                          Nothing -> False
+
+opReturnType :: Op -> PrimType
+opReturnType (OpBin  op) = binOpType op
+opReturnType (OpCmp  op) = cmpOpType op
+opReturnType (OpUn   op) = unOpType  op
+opReturnType (OpConv op) = snd $ convOpType op
+opReturnType (OpFn   fn) = fromJust $ do
+                            (_, t, _) <- M.lookup fn primFuns
+                            Just t
+
+addFor :: Op -> Op
+addFor op = OpBin $ fromJust $ case opReturnType op of
+  IntType t -> Just $ Add t OverflowUndef
+  FloatType t -> Just $ FAdd t
+  _ -> Nothing
+
+multiplyFor :: Op -> Op
+multiplyFor op = OpBin $ fromJust $ case opReturnType op of
+  IntType t -> Just $ Mul t OverflowUndef
+  FloatType t -> Just $ FMul t
+  _ -> Nothing
+
+typeAsValue :: PrimType -> Rational -> PrimValue
```

Comment:
This is almost identical to `valueAsType`. Surely some code can be deduplicated?